### PR TITLE
Export track after recording

### PR DESCRIPTION
**Pull request by thrillfall**
# Thanks for your contribution.

**Describe the pull request**
Instant post-workout gpx export to external storage

**Link to the the issue**
#186 #90 

**License agreement**
By opening this pull request, you are providing your contribution under the _Apache License 2.0_ (see [LICENSE.md](LICENSE.md)).

### CODE REVIEW

**Comment by thrillfall** [comment]
i copy&pasted a lot of the exporter code and reduced it to only work with a single track and get rid of the unused (at least in my case) dependency to ApplicationContext. This is as far as i got to make a simple gpx export work. 
Any suggestions as to how to improve this changeset are highly appreciated.

**Comment by dennisguse** [comment]
Did not read the code (phone only).
Would you mind modifying your code to support the preference (default file format), so the user can set the format.

And what is your use case?

**Comment by thrillfall** [comment]
My usecase is that i'd like to have my tracks (after finishing the workout) uploaded automatically to my nextcloud where i can then enjoy the tracks in the maps app.

I did not want to restrict the exportable format to only gpx in the beginning. But after reading the exporter related code i found that there are issues that i did not want to build on top off but rather duplicate parts of the code to make the smallest useful functionality (GPX export) working.
The biggest issues that i found was that FileTrackExporter and KmzTrackExporter share the same interface but FiletrackExporter   does not need the Context in writeTrack(). It seems both exporters have evolved in different ways but are still stuck to the old interface. Context should maybe be injected by constructor (in KmzTrackExporter) thus keeping the interface clean of only partialy used dependencies.
Secondly i did not understand at all how all the instantiations of the different exporter strategies where done in TrackFileFormat. 
An enums that yields static factory methods is something i have never seen before. But i am no Java developer so i assume its common.
Third and last i did not want to use the interface of FileTrackExporter that expects an array of tracks where i only have s single track to export. I'd assume that there are 2 classes (or at least methods) where one deals with exporting of a single track and the other of an array of tracks. All but one usage of the exporter pass an array of the single track that needs exporting. The only usage of actually more then one track to export was some "weird" import from file by splitting the filename by underscore.

TL;DR; it was hard for me to understand the code and use it as is in satisfying manner (mostly because i really could not grasp the complexity of the implementation as i am no java developer). So started with the smallest increment that i needed.

I subsequent increments i'd like to evolve this into a multiformat automatic export but i'd start out refactoring the current implementation (at least i'd try :smile:)

**Comment by dennisguse** [comment]
Yeah, this part of the code needs some love - but you got something working :+1: 
I am at the moment working on exporting/importing the UUID of a track - I will extend the original exporter/importer code.

From my perspective, it would be better to slowly refactor the old code (I know it is really ............).

One thing: (although it is not used at the moment) OpenTracks supports reading/writing multiple tracks into one GPX/KML/KMZ. I don't know what the original use case was, but I would actually like to keep this functionality. Other applications might create files using this approach. I also was not aware that this is handled as it is implemented.

Some days ago, I created a class PlantUML diagram: https://github.com/OpenTracksApp/OpenTracks/pull/312/commits/65349d4b1e2dd011c332e64c96132ff8e1102f47
This might be helpful.

One thing about functionality: I guess it would be better to make the folder to be exported configurable (via settings) instead of asking the user every time (just make sure to persist the permissions of the URL you get from the directory chooser as otherwise a device reboot will revoke these permissions). Then you can also skip the newly implemented activity.

**Comment by dennisguse** [comment]
PS: TrackFileFormat is also weird - we should do this differently.
Suggestions welcome.

**Comment by thrillfall** [comment]
> One thing about functionality: I guess it would be better to make the folder to be exported configurable (via settings) instead of asking the user every time (just make sure to persist the permissions of the URL you get from the directory chooser as otherwise a device reboot will revoke these permissions). Then you can also skip the newly implemented activity.

i thought thats what i did with src/main/java/de/dennisguse/opentracks/io/file/importer/SingleTrackExportDirectoryChooserActivity.java?
I set the target directory in the settings and once the export happens it writes the file to that directory. Did i miss anything?

**Comment by thrillfall** [comment]
I currently trying to refactor the "multi-track" import/export and see what can be simplified. Until then i'd close this MR and open a new one once i have a satisfying solution that can do all the formats.
Thanks for your esponses and interest in the contribution

**Comment by dennisguse** [comment]
@thrillfall You are right, I misread SingleTrackExportDirectoryChooserActivity.
You are storing it as preference, but you also need to store the permission that you can write to this directory later on (after device restart).
See https://developer.android.com/training/data-storage/shared/documents-files
and search for persist permissions.
Also I think this preference should be reachable via the SettingsActivity as well.

**Comment by dennisguse** [comment]
@thrillfall If you have any questions, let us know.
You can also open an issue for this.
